otel_instrumentation_mcp/cache.py

```diff
@@ -0,0 +1,588 @@
+                        },
+                    )
+                else:
+                    logger.warning(
+                        "Failed to cache data",
+                        extra={
+                            "operation": operation,
+                            "cache_key": cache_key,
+                        },
+                    )
+
+                return result
+
+            except Exception as e:
+                set_span_error(span, e)
+                logger.error(
+                    "Cache operation failed",
+                    exc_info=True,
+                    extra={
+                        "operation": operation,
+                        "cache_key": cache_key,
+                    },
+                )
+                # If caching fails, still try to fetch the data
+                return await fetch_func()
+
+    async def invalidate(
+        self,
+        operation: str,
+        language: Optional[str] = None,
+        version: Optional[str] = None,
+        category: Optional[str] = None,
+        repo: Optional[str] = None,
+        keywords: Optional[str] = None,
+        **kwargs,
+    ) -> bool:
+        """Invalidate a specific cache entry."""
+        cache_key = self._generate_cache_key(
+            operation=operation,
+            language=language,
+            version=version,
+            category=category,
+            repo=repo,
+            keywords=keywords,
+            **kwargs,
+        )
+
+        with tracer.start_as_current_span("cache.invalidate") as span:
+            add_span_attributes(
+                span,
+                **{
+                    SpanAttributes.CODE_FUNCTION: "invalidate",
+                    "cache.operation": operation,
+                    "cache.key": cache_key,
+                },
+            )
+
+            try:
+                result = await self.backend.delete(cache_key)
+                span.add_event(
+                    "cache_invalidated",
+                    {
+                        "cache.key": cache_key,
+                        "cache.success": result,
+                    },
+                )
+                return result
+            except Exception as e:
+                set_span_error(span, e)
+                logger.error(
+                    "Cache invalidation failed",
+                    exc_info=True,
+                    extra={"cache_key": cache_key},
+                )
+                return False
+
+    async def health_check(self) -> Dict[str, Any]:
+        """Check cache health."""
+        return await self.backend.health_check()
+
+
+def create_cache_manager() -> CacheManager:
+    """Create cache manager based on configuration."""
+    cache_enabled = os.getenv("CACHE_ENABLED", "false").lower() == "true"
+    cache_backend = os.getenv("CACHE_BACKEND", "memory").lower()
+
+    if not cache_enabled:
+        logger.info("Caching is disabled, using in-memory cache with short TTL")
```

Comment:
Curious why we still keep a short cache even when caching is disabled? Can see potential for user confusion if they think caching is fully disabled, but still get cache hits within a short time frame